### ISSUE TITLE
Incorrect error messages of incompatible method errors

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#6177]
```ts
class Base {
	private base;
	m() {
	}
}
class Derived extends Base {
	private derived;
	m(){}
}

<Derived>new Base(); // ok
```

```ts
class Base {
	private base;
	m() {
	}
}
class Derived extends Base { // error
	private derived;
	m(a){} // change a method signature
}

<Derived>new Base(); // error
```

```
$ node built/local/tsc.js -m commonjs index --noEmit
index.ts(6,7): error TS2415: Class 'Derived' incorrectly extends base class 'Bas
e'.
  Types of property 'm' are incompatible.
    Type '(a: any) => void' is not assignable to type '() => void'.
index.ts(11,1): error TS2352: Neither type 'Base' nor type 'Derived' is assignab
le to the other.
  Property 'derived' is missing in type 'Base'.
```

A second error message is wrong. Should be same as a first message.

**Comment by DanielRosenwasser** [comment]
We check assignability in both directions, and neither error is wrong in a sense, so I don't think this is a bug.

**Comment by falsandtru** [comment]
When I get that error, I think that I should fix properties. But really correction target is methods, not properties. This error message mislead the correction target. This error message is lying. I think that this error message should be corrected.

**Comment by RyanCavanaugh** [comment]
A method *is* a property.

**Comment by Arnavion** [comment]
I believe @falsandtru 's point is that since the incompatibility of `m`'s signature is the only thing preventing the type assertion from succeeding, the error message should be about that, instead of saying `derived` doesn't exist in Base (`derived` doesn't have to exist in Base anyway since Derived is derived from Base, as the first example shows).

**Comment by RyanCavanaugh** [comment]
I don't understand how we would even do that. For any given failed type assertion `<T>u`, there are two different error messages we could give -- an error about `u` being unassignable to `T`, or an error about `T` being unassignable to `u`.

There's always going to be some root cause for both incompatibilities. There's no plausible way to take two incompatibilities and decide which is more likely to be the one you consider to be more relevant. That would literally require reading the user's mind.

**Comment by Arnavion** [comment]
The fact that type assertion checks assignability both ways isn't relevant. The same situation manifests with `let d: Derived = new Base();` That is, for your example, the error is only about assignability of U to T.

I haven't looked at the checker, but I'm assuming what's going on is what I described in the edit of my previous comment.

**Comment by RyanCavanaugh** [comment]
In this case

```ts
class Base {
    private base;
    m() {
    }
}
class Derived extends Base { // error
    private derived;
    m(a){} // change a method signature
}

let d: Derived = new Base();
```

the order of operations is

 * Check that `Derived` is assignable to `Base` because of the `extends` clause
  * Check that `m`'s signature is assignable
   * It's not; issue an error
 * Check that `Base` is assignable to `Derived` in the `let` statement
  * Check that `m`'s signature is assignable
    * It is
  * Check that `Base` has the properties of `Derived`
    * It doesn't, issue an error

Every single one of those steps is correct. In the `let` statement we're checking the opposite relation (is `Base` assignable to `Derived`) to what we check in the `extends` clause (is `Derived` assignable to `Base`).

In the "check `Base` assignable to `Derived`" step, there is nothing wrong with `m` and nothing to complain about.

**Comment by falsandtru** [comment]
> There's no plausible way to take two incompatibilities and decide which is more likely to be the one you consider to be more relevant.

I understand, then we can tell a more accurate message by telling about other ways. How about?

**Comment by Arnavion** [comment]
@RyanCavanaugh I fixed a typo shortly after I posted it. I meant to say `let b: Base = new Derived();`

You can see this one gives an error about incompatible `m`. Ideally when checking `<Derived>new Base()` the checker would see that Derived must be assignable to Base due to the inheritance relationship, and report errors only about why that relationship is not valid, just like `let b: Base = new Derived();` does. Errors from the reverse (Base is not assignable to Derived because of missing `derived`) should be suppressed or have lower priority.

**Comment by RyanCavanaugh** [comment]
Our top-level assumption when assessing error quality is that people read error messages top-down. This is manifest everywhere -- due to caching, for example, you're only going to get elaborated error messages the first time a particular relation fails. So if `Derived` isn't assignable to `Base`, that's going to be an error at the declaration site, which should come before any usage. 

As for special casing which error to report when checking assignability for the type assertion, I think it's probably a fool's errand. If the derived type isn't compatible with the base type, your program is already not compiling and needs help elsewhere. Maybe someone can send a PR that offers up convincing improvements, but I don't think this actually comes up with any frequency that would merit complicating that code.

**Comment by Arnavion** [comment]
Yes, I understand. I can imagine a situation where the two classes each and the assertion are far apart (say three separate files) and the user isn't reading errors top-to-bottom, but it does seem quite a stretch and rare...
